frontend/catalyst/api_extensions/control_flow.py

```diff
@@ -283,6 +283,11 @@ def for_loop(lower_bound, upper_bound, step, loop_fn, *args):
         lower_bound (int): starting value of the iteration index
         upper_bound (int): (exclusive) upper bound of the iteration index
         step (int): increment applied to the iteration index at the end of each iteration
+        allow_array_resizing (bool): Whether to allow arrays to change shape/size within
+            the for loop. By default this is ``False``; this will allow out-of-scope
+            dynamical-shaped arrays to be captured by the for loop. Set this to ``True``
+            to modify array shape/size within the for loop, however outer-scope
+            dynamical-shaped arrays will no longer be captured.
```

Comment:
I think we should also talk about the other effect this flag has (why it was originally introduced, although the closure might be even more important).

Just confirmed that setting the flag to true prevents two arrays with same shape from being used in a binary operator together:
```py
@qjit
def f(n: int, m: int):
    x = jnp.ones((n, m), dtype=float)
    y = jnp.ones((n, m), dtype=float)

    @for_loop(0, n, 1, experimental_preserve_dimensions=True)  # does not work if False
    def sum_and_multiply(i, x, y):
        x = x + y
        y = x * y
        return x, y

    return sum_and_multiply(x, y)

f(2, 3)
```

Reply:
Perfect, thanks david! I was missing an example of this case and couldn't figure out how to recreate it myself.